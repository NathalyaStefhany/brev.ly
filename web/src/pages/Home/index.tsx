import React from 'react';

import { NewLink } from '@/pages/Home/components/NewLink';
import logo from '@/assets/logo.svg';

export const Home: React.FC = () => {
  return (
    <main className="min-h-screen flex flex-col items-center py-16 px-6 bg-gray-200">
      <div className="flex flex-col items-start gap-16">
        <img
          src={logo}
          alt="Ícone de corrente azul representando um link ao lado do nome 'brev.ly' também em azul."
          className="h-12"
        />

        <div className="flex flex-row flex-wrap justify-center gap-10">
          <NewLink />
        </div>
      </div>
    </main>
  );
};
