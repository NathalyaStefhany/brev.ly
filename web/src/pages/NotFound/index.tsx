import React from 'react';

import notFoundImage from '@/assets/404.svg';

export const NotFound: React.FC = () => {
  return (
    <main className="w-screen min-h-screen flex flex-row items-center justify-center p-6 bg-gray-200">
      <div className="max-w-290 w-fit h-fit flex flex-col gap-12 items-center text-center bg-gray-100 rounded-lg px-24 py-32">
        <img
          src={notFoundImage}
          alt="Número 404 representando o erro retornado ao acessar a página"
          className="h-42.5"
        />

        <h1 className="text-xl text-gray-600">Link não encontrado</h1>

        <p className="text-md text-gray-500">
          O link que você está tentando acessar não existe, foi removido ou é
          uma URL inválida. Saiba mais em{' '}
          <a href="/" className="text-blue-base underline">
            brev.ly
          </a>
          .
        </p>
      </div>
    </main>
  );
};
