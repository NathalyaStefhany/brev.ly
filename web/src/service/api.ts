import { env } from '@/env';
import axios from 'axios';

const baseURL = env.VITE_API_URL;

export const api = axios.create({
  baseURL,
});
