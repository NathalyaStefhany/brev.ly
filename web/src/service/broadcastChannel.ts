export const refetchChannel = new BroadcastChannel('refetch-channel');
