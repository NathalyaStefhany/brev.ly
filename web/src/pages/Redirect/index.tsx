import React, { useEffect, useState } from 'react';

import { useLocation, useNavigate } from 'react-router';
import { api } from '@/service/api';
import logoIcon from '@/assets/logo-icon.svg';
import { AxiosError } from 'axios';
import { queryClient } from '@/service/queryClient';
import { refetchChannel } from '@/service/broadcastChannel';

type GetOriginalLinkOutput = {
  originalLink: string;
};

export const Redirect: React.FC = () => {
  const [rendered, setRendered] = useState(false);

  const { pathname } = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    const getOriginalLink = async (): Promise<void> => {
      try {
        const { data } = await api.get<GetOriginalLinkOutput>(
          `/shortened-links/${pathname.slice(1)}/original-link`,
        );

        let url = data.originalLink;

        if (url.startsWith('www.')) {
          url = `https://${url}`;
        }

        window.location.href = url;
      } catch (error) {
        const err = error as AxiosError;

        if (err.status === 404) {
          navigate('/url/not-found');
        }
      }
    };

    const updateAccessQuantity = async (): Promise<void> => {
      try {
        await api.patch(`/shortened-links/${pathname.slice(1)}/access`);

        queryClient.refetchQueries({ queryKey: ['links list'], exact: true });

        refetchChannel.postMessage({ queryKey: ['links list'], exact: true });
      } catch (error) {
        const err = error as AxiosError;

        if (err.status === 404) {
          navigate('/url/not-found');
        }
      }
    };

    if (pathname && rendered) {
      getOriginalLink();
      updateAccessQuantity();
    } else {
      setRendered(true);
    }
  }, [pathname, navigate, rendered]);

  return (
    <main
      className="w-screen min-h-screen flex flex-row items-center justify-center p-6 bg-gray-200"
      data-testid="container-redirect-page"
    >
      <div className="max-w-290 w-full h-fit flex flex-col gap-12 items-center text-center bg-gray-100 rounded-lg px-10 py-24 md:px-24 md:py-32">
        <img
          src={logoIcon}
          alt="Logotipo do site com um ícone de corrente azul que representa um link."
          className="h-24"
        />

        <h1 className="text-xl text-gray-600">Redirecionando...</h1>

        <div className="flex flex-col gap-2">
          <p className="text-md text-gray-500">
            O link será aberto automaticamente em alguns instantes.
          </p>

          <p className="text-md text-gray-500">
            Não foi redirecionado?{' '}
            <a
              href="/"
              className="text-blue-base underline"
              data-testid="link-access-here"
            >
              Acesse aqui
            </a>
          </p>
        </div>
      </div>
    </main>
  );
};
