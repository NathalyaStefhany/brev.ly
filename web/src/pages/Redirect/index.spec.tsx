import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import AxiosMock from 'axios-mock-adapter';
import { Redirect } from '@/pages/Redirect';
import { api } from '@/service/api';
import { queryClient } from '@/service/queryClient';

const mockNavigate = vi.fn();

vi.mock('react-router', async () => {
  const actual = await vi.importActual('react-router');
  return {
    ...actual,
    useLocation: (): { pathname: string } => ({
      pathname: '/test',
    }),
    useNavigate: (): (() => void) => mockNavigate,
  };
});

const apiMock = new AxiosMock(api);

describe('Redirect page tests', () => {
  beforeEach(() => {
    apiMock.reset();
  });

  it('should render correctly', () => {
    const { container } = render(<Redirect />);

    expect(container).toMatchSnapshot();
  });

  it('should go to home page when click on "Access here" link', () => {
    render(<Redirect />);

    expect(screen.getByTestId('link-access-here')).toHaveAttribute('href', '/');
  });

  it('should redirect to the original page when url starts with "www."', async () => {
    apiMock
      .onGet('/shortened-links/test/original-link')
      .reply(200, { originalLink: 'www.test.com' });

    const locationMock = { href: '' };

    Object.defineProperty(window, 'location', {
      value: locationMock,
      writable: true,
    });

    render(<Redirect />);

    await waitFor(() => {
      expect(
        apiMock.history.get.some(
          ({ url }) => url === '/shortened-links/test/original-link',
        ),
      );
    });

    expect(locationMock.href).toBe('https://www.test.com');
  });

  it('should redirect to the original page when url does not start with "www."', async () => {
    apiMock
      .onGet('/shortened-links/test/original-link')
      .reply(200, { originalLink: 'https://www.test.com' });

    const locationMock = { href: '' };

    Object.defineProperty(window, 'location', {
      value: locationMock,
      writable: true,
    });

    render(<Redirect />);

    await waitFor(() => {
      expect(
        apiMock.history.get.some(
          ({ url }) => url === '/shortened-links/test/original-link',
        ),
      );
    });

    expect(locationMock.href).toBe('https://www.test.com');
  });

  it('should redirect to the not found page when shortened link does not exist', async () => {
    apiMock.onGet('/shortened-links/test/original-link').reply(404);

    render(<Redirect />);

    await waitFor(() => {
      expect(
        apiMock.history.get.some(
          ({ url }) => url === '/shortened-links/test/original-link',
        ),
      );
    });

    expect(mockNavigate).toBeCalledWith('/url/not-found');
  });

  it('should update access quantity when access the page', async () => {
    apiMock.onPatch('/shortened-links/test/access').reply(200);

    queryClient.refetchQueries = vi.fn();

    render(<Redirect />);

    await waitFor(() => {
      expect(
        apiMock.history.patch.some(
          ({ url }) => url === '/shortened-links/test/access',
        ),
      );
    });

    expect(queryClient.refetchQueries).toBeCalledWith({
      queryKey: ['links list'],
      exact: true,
    });
  });
});
