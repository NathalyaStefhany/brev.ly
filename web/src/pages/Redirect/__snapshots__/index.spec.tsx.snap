// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Redirect page tests > should render correctly 1`] = `
<div>
  <main
    class="w-screen min-h-screen flex flex-row items-center justify-center p-6 bg-gray-200"
    data-testid="container-redirect-page"
  >
    <div
      class="max-w-290 w-full h-fit flex flex-col gap-12 items-center text-center bg-gray-100 rounded-lg px-10 py-24 md:px-24 md:py-32"
    >
      <img
        alt="Logotipo do site com um ícone de corrente azul que representa um link."
        class="h-24"
        src="data:image/svg+xml,%3csvg%20width='52'%20height='52'%20viewBox='0%200%2052%2052'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20fill-rule='evenodd'%20clip-rule='evenodd'%20d='M43.4044%2016.8233C47.2925%2019.8534%2049.2083%2024.0438%2049.2782%2028.8694C49.3244%2032.0642%2048.7605%2035.2044%2047.3133%2037.9635C46.5857%2039.2025%2045.6672%2040.3565%2044.532%2041.3927C40.8665%2044.7386%2035.9478%2045.9247%2030.9245%2045.9966C25.9013%2046.0684%2020.9503%2045.0233%2017.1894%2041.7837C15.5286%2040.3531%2014.302%2038.6717%2013.4357%2036.8332C11.5802%2036.9439%209.67833%2036.4069%208.10056%2035.1784L8.09841%2035.1767C4.21032%2032.1466%202.29446%2027.9562%202.2246%2023.1307C2.17835%2019.9358%202.7426%2016.795%204.18987%2014.0359C4.91744%2012.7969%205.83563%2011.6435%206.97083%2010.6073C10.6363%207.26145%2015.555%206.07526%2020.5783%206.00343C25.6015%205.9316%2030.5525%206.97666%2034.3134%2010.2163C35.9741%2011.6469%2037.2008%2013.3283%2038.0671%2015.1669C39.9226%2015.0561%2041.8245%2015.5931%2043.4022%2016.8216L43.4044%2016.8233ZM10.6611%2022.8958C10.6611%2025.1315%2011.3636%2026.6969%2012.9402%2027.9727C13.6477%2028.5453%2013.7598%2029.5894%2013.1876%2030.301C12.6153%2031.0125%2011.578%2031.1234%2010.8688%2030.5492C8.51054%2028.6379%207.36272%2026.1359%207.36272%2022.8958C7.36272%2015.0406%2011.8089%2011.0575%2020.5841%2011.0575C29.3593%2011.0575%2033.8071%2015.0406%2033.8071%2022.8958C33.8071%2026.1458%2032.6593%2028.6495%2030.2977%2030.5509C29.9942%2030.7958%2029.6298%2030.9149%2029.2669%2030.9149C28.7854%2030.9149%2028.3055%2030.7031%2027.9806%2030.296C27.41%2029.5828%2027.5238%2028.5403%2028.2346%2027.9694C29.8079%2026.7018%2030.5104%2025.1364%2030.5104%2022.8958C30.5104%2018.6778%2029.3313%2014.3671%2020.5857%2014.3671C11.8402%2014.3671%2010.6611%2018.6795%2010.6611%2022.8958ZM37.8195%2021.6018C38.3918%2020.8919%2039.4291%2020.7777%2040.1382%2021.3536V21.3519C42.4966%2023.2632%2043.6444%2025.7652%2043.6444%2029.0053C43.6444%2036.8605%2039.1966%2040.8436%2030.4214%2040.8436C21.6461%2040.8436%2017.1983%2036.8605%2017.1983%2029.0053C17.1983%2025.7553%2018.3462%2023.2516%2020.7078%2021.3503C21.4169%2020.7777%2022.4559%2020.8902%2023.0265%2021.6051C23.5971%2022.3183%2023.4833%2023.3608%2022.7725%2023.9317C21.1992%2025.1993%2020.4967%2026.7647%2020.4967%2029.0053C20.4967%2033.2233%2021.6758%2037.534%2030.4214%2037.534C39.1669%2037.534%2040.346%2033.2217%2040.346%2029.0053C40.346%2026.7697%2039.6435%2025.2043%2038.0669%2023.9284C37.3594%2023.3559%2037.2473%2022.3134%2037.8195%2021.6018Z'%20fill='%232C46B1'/%3e%3c/svg%3e"
      />
      <h1
        class="text-xl text-gray-600"
      >
        Redirecionando...
      </h1>
      <div
        class="flex flex-col gap-2"
      >
        <p
          class="text-md text-gray-500"
        >
          O link será aberto automaticamente em alguns instantes.
        </p>
        <p
          class="text-md text-gray-500"
        >
          Não foi redirecionado?
           
          <a
            class="text-blue-base underline"
            data-testid="link-access-here"
            href="/"
          >
            Acesse aqui
          </a>
        </p>
      </div>
    </div>
  </main>
</div>
`;
